subsys/fs/fs.c

```diff
@@ -721,6 +733,51 @@ int fs_mount(struct fs_mount_t *mp)
 	return rc;
 }
 
+#if defined(CONFIG_FILE_SYSTEM_MKFS)
+
+int fs_mkfs(struct fs_mount_t *mp)
+{
+	const struct fs_file_system_t *fs;
+	int rc = -EINVAL;
+
+	k_mutex_lock(&mutex, K_FOREVER);
+
+	rc = validate_mp(mp, NULL);
+	if (rc < 0) {
+		goto mount_err;
+	}
+
+	/* Get file system information */
+	fs = fs_type_get(mp->type);
+	if (fs == NULL) {
+		LOG_ERR("requested file system type not registered!!");
+		rc = -ENOENT;
+		goto mount_err;
+	}
```

Comment:
This part can also be taken under `validate_mp`, as this is the same code as in `fs_mount`.
Can you also change the message to print the inf code for the type? I think it is missing that information.

Reply:
I don't think that getting file system information should be a part of `validate_mp`. It is very different operation then checking if mount point is valid. Functions that do one thing are easier to understand hence I don't want to do anything other then checking mount point in that function. I suggest to leave `fs_type_get` where it is.

Reply:
I have added information about a mp type to the error message.

Reply:
> I don't think that getting file system information should be a part of `validate_mp`. It is very different operation then checking if mount point is valid. Functions that do one thing are easier to understand hence I don't want to do anything other then checking mount point in that function. I suggest to leave `fs_type_get` where it is.

Is mount point object valid when  `mp->type` is ininvalid?

The  `validate_mp` only validates `mp` object in context of Zephyr subsystem, you are still not warranted to mount the file system, because, for example, FAT FS will only mount at specified mount points.
https://github.com/zephyrproject-rtos/fatfs/blob/master/include/ffconf.h#L204
This is probably thing for the next PR...

Reply:
The purpose of the `validate_mp` function is to check if mp is valid in the sense of Zephyr subsystem. We cannot check if mp is valid for FAT FS if we don’t have retrieved file system information.

The file system type may be an arbitrary number (see: https://github.com/zephyrproject-rtos/zephyr/blob/07de2ffbb01a4633cdec7e48190ae369ecc37c9b/include/zephyr/fs/fs.h#L42:L47). Then it is impossible to check if that field is valid without trying to access it.

Maybe I will describe it in an analogy to the file system. One thing is to check if the file path is valid (consists of allowed characters). The other thing is looking for a file in the file system. If a file is not found then it doesn’t mean that path was invalid. It just means that file does not exist.

If you still think that `fs_type_get` should be inside `validate_mp` I will add another argument to that function that will be used to return the fs information.

Reply:
> The purpose of the `validate_mp` function is to check if mp is valid in the sense of Zephyr subsystem. We cannot check if mp is valid for FAT FS if we don’t have retrieved file system information.

You did not get my point. I do not expect you to validate mount point for the FAT FS or any other system. What you check here is validity of mp object in Zephyr context.

> 
> The file system type may be an arbitrary number (see: https://github.com/zephyrproject-rtos/zephyr/blob/07de2ffbb01a4633cdec7e48190ae369ecc37c9b/include/zephyr/fs/fs.h#L42:L47). Then it is impossible to check if that field is valid without trying to access it.

This argument makes no sense.  All systems are registered with `fs_register`, that includes Zephyr supported FAT and LittleFS. User may `fs_register` another FAT driver with new identifier (`fs_type`) and use it with `fs_mount_t` object definition.
VFS subsystem holds own list of know fs types and `fs_type_get` searches the list for that type.
If `validate_mp` checks mount point object, in the Zephyr context, it can also check whether the object carries valid, already registered type.

> 
> Maybe I will describe it in an analogy to the file system. One thing is to check if the file path is valid (consists of allowed characters). The other thing is looking for a file in the file system. If a file is not found then it doesn’t mean that path was invalid. It just means that file does not exist.

OK, so we have different understanding of `validate_mp`: I understand it as validation of `mp` object, as object defining mount point with valid path, not duplicating already mounted path, containing containing valid valid `fs_type`, which is valid candidate for being added to `fs_mnt_list`, if successfully mounted by the given file system type.

> 
> If you still think that `fs_type_get` should be inside `validate_mp` I will add another argument to that function that will be used to return the fs information.

Yes, that was what I have been asking for.

Now, after all the arguing for the change I actually think that we wen the wrong way from the beginning, and that signature of `fs_mkfs` should be:
`int fs_mkfs(int fs_type, uintptr dev_id)`

The `fs_mkfs` should only process fs_type and maybe check whether dev_id is not NULL, all other processing should be dumped on mkfs specific for file system.

Reply:
Good point. I will try to make the new signature for that function and it may became simpler.

Reply:
> Good point. I will try to make the new signature for that function and it may became simpler.

Wait, there is one problem with this: we do not handle records of "used devices", which means that you could mkfs on device that is mouted, something that your existing solution would not allow!